adapters/insticator/insticator.go

```diff
@@ -0,0 +1,324 @@
+	if response.Cur != "" {
+		bidResponse.Currency = response.Cur
+	}
+	for _, seatBid := range response.SeatBid {
+		for i := range seatBid.Bid {
+			bid := &seatBid.Bid[i]
+			bidType := getMediaTypeForBid(bid)
+			b := &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: bidType,
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+	return bidResponse, nil
+}
+
+func makeImps(imp openrtb2.Imp) (openrtb2.Imp, string, string, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := jsonutil.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return openrtb2.Imp{}, "", "", &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var insticatorExt openrtb_ext.ExtImpInsticator
+	if err := jsonutil.Unmarshal(bidderExt.Bidder, &insticatorExt); err != nil {
+		return openrtb2.Imp{}, "", "", &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	// Directly construct the impExt
+	impExt := ext{
+		Insticator: impInsticatorExt{
+			AdUnitId:    insticatorExt.AdUnitId,
+			PublisherId: insticatorExt.PublisherId,
+		},
+	}
+
+	impExtJSON, err := jsonutil.Marshal(impExt)
+	if err != nil {
+		return openrtb2.Imp{}, "", "", &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+	imp.Ext = impExtJSON
+
+	// Validate Video if it exists
+	if imp.Video != nil {
+		if err := validateVideoParams(imp.Video); err != nil {
+			return openrtb2.Imp{}, insticatorExt.AdUnitId, insticatorExt.PublisherId, &errortypes.BadInput{
+				Message: err.Error(),
+			}
+		}
+	}
+
+	// Return the imp, AdUnitId, and no error
+	return imp, insticatorExt.AdUnitId, insticatorExt.PublisherId, nil
+}
+
+func makeReqExt(request *openrtb2.BidRequest) ([]byte, error) {
+	var reqExt reqExt
+
+	if len(request.Ext) > 0 {
+		if err := jsonutil.Unmarshal(request.Ext, &reqExt); err != nil {
+			return nil, err
+		}
+	}
+
+	if reqExt.Insticator == nil {
+		reqExt.Insticator = &reqInsticatorExt{}
+	}
+
+	if reqExt.Insticator.Caller == nil {
+		reqExt.Insticator.Caller = make([]insticatorCaller, 0)
+	}
+
+	reqExt.Insticator.Caller = append(reqExt.Insticator.Caller, caller)
+
+	return jsonutil.Marshal(reqExt)
+}
+
+func resolveBidFloor(bidFloor float64, bidFloorCur string, reqInfo *adapters.ExtraRequestInfo) (float64, error) {
+	if bidFloor > 0 && bidFloorCur != "" && strings.ToUpper(bidFloorCur) != "USD" {
+		floor, err := reqInfo.ConvertCurrency(bidFloor, bidFloorCur, "USD")
+		return mathutil.RoundTo4Decimals(floor), err
+	}
+
+	return bidFloor, nil
+}
+
+func validateVideoParams(video *openrtb2.Video) error {
+	if (video.W == nil || *video.W == 0) || (video.H == nil || *video.H == 0) || video.MIMEs == nil {
+		return &errortypes.BadInput{
+			Message: "One or more invalid or missing video field(s) w, h, mimes",
+		}
+	}
+
+	return nil
+}
+
+// populatePublisherId function populates site.publisher.id or app.publisher.id
+func populatePublisherId(publisherId string, request *openrtb2.BidRequest) error {
```

Comment:
This function does not return an error anymore so please remove that return parameter.